Pv generator

pford opened the pull request:
Closes #794.

The implementation is done in RegionHandler, which approximates the line as a series of box regions with the requested width.  The PV image data is the spectral profile of each box region set in an array.  This code can be reused for the mean spatial profile along a line by getting the mean of each region for the current channel instead of per-channel.  Polyline is not currently supported.

For images with fixed pixel separation, a series of temporary rectangle regions is created along the line.  If the angular separation of the fixed pixel points varies, a fixed angular separation (variable pixel interval) is used to create a series of temporary polygon regions to describe a box with four corners.  This is because the pixel center for a rectangle may not be the angular center.  The mean of three adjacent boxes was used for the angular separation to avoid a region with no pixel centers, resulting in a nan spectral profile.

Each rectangle or polygon region is set with a temporary region id, then its spectral profile is calculated and added as a row in the 2D array which will be used for the PV image data.  Then the temporary region is destroyed.

Once the data array is set, it is passed to the PvGenerator which creates an image with the appropriate CoordinateSystem and sets the data.  The x-axis is an Offset axis with offset 0 at the line center, positive offsets toward endpoint 0, and negative offsets toward endpoint 1.  This image is passed back to Session which opens it, similar to the MomentGenerator.

I created a directory ImageGenerators for the MomentGenerator and PvGenerator with a shared header ImageGenerator.h.  The CollapseResult struct has been generalized to a GeneratedImage struct in this header file for the MomentGenerator and PvGenerator.

Region implementation document:
https://docs.google.com/document/d/1VZqXM-A8Djmsybz4kNJsXudoghnjtqFvadB8SxlqEUQ/edit#

pford committed:
Add pv protobuf messages

pford committed:
Update protobuf with pv messages fixes

pford committed:
Change directory name for image generators

pford committed:
Add handlers for PV messages

pford committed:
Add common generator definitions

pford committed:
Merge branch 'dev' into pv_generator

pford committed:
Add framework for pv generator

pford committed:
Define interfaces for pv image generation functions

pford committed:
Merge branch 'dev' into pv_generator

pford committed:
Complete PvGenerator for per-channel stats/ per-region spectral calc

pford committed:
Call OnPvRequest from OnSetRegion for testing only

pford committed:
Pv generator for uniform pixel increment only

For testing; not quite correct yet, but seems to produce an image.

pford committed:
Merge branch 'dev' into pv_generator

pford committed:
Fix Moment path

pford committed:
Fix order of includes

pford committed:
Use LCRegion instead of ImageRegion for PV generator

pford committed:
Refactor pv gen stats/spectral calc into PvGenerator class

pford committed:
Fixes for PvGenerator TempImage

pford committed:
Refactor pixel/angular regions for line approximation

pford committed:
Add performance logging to pv image calculation

pford committed:
Move pv performance logging to Session to include line regions

pford committed:
Add Pv messages to SessionManager

pford committed:
Pv calculation fails when no spectral axis

pford committed:
Return after error in PvGenerator

pford committed:
Calculate fixed angular separation for box regions along pv cut

pford committed:
Use file system header

pford committed:
Use StatsCalculator for all channels per region in pv generator

pford committed:
Fix formatting

kswang1029 commented:
@pford for a cube with a pixel size of 0.05 arcsec, the offset pixel of the pv image generated with carta is 0.0494949". Could you please check possible places for the error here?

I also noticed that given the same line region definition as from-to (eg [100, 100] to [200, 100]) in image coordinate, the offset axis has two less pixels (eg 99) comparing to the pv image generated with impv (eg 101). Could you check too?

kswang1029 commented:
@pford When a pv cut is partially outside the image or totally outside the image, we need a proper handle of these cases. For the partially outside case, we should still be able to generate a pv image with those outside part as NaN. For the totally outside case, the backend should provide a proper error message to the frontend (and maybe the frontend can check if the line region and the image have intersection before enabling the generate button (@TienHao)
.

kswang1029 commented:
@pford are you able to generate a pv image if the input is a cube with IQUV as Stokes? I see it runs with few progress update messages and stops. The backend log shows 

> TBB Warning: Exact exception propagation is requested by application but the linked library is built without support for it

Not sure if this is due to my build env or else. (adding @ajm-asiaa in the loop)

pford committed:
Fix pv generator progress update

pford committed:
Fix increment calculation for pv generator

pford committed:
Set pv image restoring beam and complete pv gen progress

pford commented:
@kswang1029 Several fixes have been added:

- Fix calculation of increment
- Update progress for null regions and when complete
- When input image has a beam set, use first beam to set pv image restoring beam (as done in imageanalysis)

When the PV generator fails (response success=false with message), this error is not displayed by the frontend, which is confusing when nothing happens.  I sent a message to @TienHao .

I tested the pv generator for two IQUV images:

- The first image failed because it had no spectral coordinate, so the backend sent an error message in the response, "Cannot generate PV image with no valid spectral axis".  Casacore failed to create the spectral coordinate from the headers since its axis CTYPE was "Frequency" instead of "FREQ".  Should I add a workaround for this?  The check was added for e.g. 2D images. And normally, the PV image spectral coordinate is simply a copy of the input image spectral coordinate.
- The second image succeeded after the beam set fix.  Before the fix, this issue caused the backend to hang.
- I did not see the TBB warning.

pford commented:
@kswang1029 more detail on pv cuts outside the image: the line box regions outside the image are null.  When the box regions vector ended in null regions, the progress was never updated to completion, so the PV image was not created. The error "PV generator failed to complete" was sent but not displayed.  This has been fixed so that, as you indicated, the null regions produce NaN data in the generated PV image.

When ALL regions are null, the error message "PV calculation failed for input region" is sent.  This message may be too vague (and poorly worded), but I did not know if there may be another case where all regions are null besides being outside the image (?) so I did not want to be too specific.

kswang1029 commented:
@pford regarding the TBB warning (and backend hangs) issue, could you try the attached image and a cut like the screenshot? In my test env, the TBB warning shows up and the backend hangs.

![Screen Shot 2021-10-13 at 3 09 03 PM](https://user-images.githubusercontent.com/20819712/137084230-18305250-e030-4ee9-a156-a189e3c5cb43.png)

https://drive.google.com/file/d/1oVt7awaUOk1tebmuaIaBPwY7crM7zY7Y/view?usp=sharing

update: we need to consider to following cases
![SDC335 579-0 292 spw0 line fits-image-2021-10-13-15-44-48](https://user-images.githubusercontent.com/20819712/137089681-98e32acf-dab9-40aa-bf9c-04f51514ab06.png)

pford committed:
Round up number of offsets (half of pixel length of cut)

pford committed:
Improve error message for no line box regions

pford committed:
Check for empty spectral profile for all NAN values in region

pford commented:
@kswang1029 I did not get the TBB message but the pv generator was hanging when trying to assign the spectral profile to the PV image data array row.  It turns out the profile is empty when all input values are NAN.  Fixed.

kswang1029 commented:
> @kswang1029 I did not get the TBB message but the pv generator was hanging when trying to assign the spectral profile to the PV image data array row. It turns out the profile is empty when all input values are NAN. Fixed.

YES! Now the hanging issue is gone. I will do more tests and report back.

kswang1029 commented:
@pford @TienHao 
I found two issues:
1) If I request a cancellation once, I cannot request a new generation anymore.
2) If I have two cubes matched and if we use the GUI to request a pv generation from the 2nd image, the backend will hang. (and I see the TBB warning - this is probably just my build env problem).

kswang1029 commented:
@pford @TienHao I found two issues:

 1. If I request a cancellation once, I cannot request a new generation anymore.
 2. If I have two cubes matched and if we use the GUI to request a pv generation from the _2nd_ image, the backend will hang. (and I see the TBB warning - this is probably just my build env problem).

kswang1029 commented:
@pford good news to report:
In order to test the case when the input cube is a wide field image, I made a test cube with a set of sources sitting at the celestial equator with _equal_ angular separation. Duo to SIN projection, those sources have variable separations on the raster image as you can see on the left panel. Then I made a pv cut alone the equator and request the pv generation. If the line box generation with equal angular separation works, we should see sources appear with equal separation on the OFFSET axis, and YES, I think they appear at right locations as we can see on the right panel. This is GREAT!.

![Screen Shot 2021-10-15 at 1 41 07 PM](https://user-images.githubusercontent.com/20819712/137438080-8d6c4ade-2f14-47a9-8671-107a4e5e6cb3.png)

PS. ~~we see some line boxes results in all NaN near the image edge, probable due to sampling issue and projection effect. This should be okay I think as I am testing an edge use case.~~
hmm for those NaN stripes, do you still use 1px as the width of the line boxes? If so I think we need to refine it to make the width be constant angular width. An approximation would be using the middle point of adjacent line box centers to build up the width. And about height, we will need to have an adaptive one. That is, calculate the angular size of 3px from header, then use the value at different places to get the height in pixel (eg 3.5px). This way, each box should approximately have equal sky coverage (aka solid angle). 

PS2: @pford @TienHao do you think we can have an option for the unit of the OFFSET axis? We should consider arcsec, arcmin, and degree. (if this will take significant extra time, we can leave it for now)

kswang1029 commented:
@pford how do we determine when we trigger "fixed angular increment" sampling? Is it something like
![Screen Shot 2021-10-15 at 2 19 08 PM](https://user-images.githubusercontent.com/20819712/137441268-a8277115-7648-4620-a0d9-b463f5ea126e.png)

pford committed:
Fix reading record for matched-image line region

pford committed:
Improve pv generator cancellation

pford committed:
Update carta-protobuf with dev

pford commented:
I fixed the matched-image pv generator issue, and improved cancellation.  It is possible that the dialog was hanging after cancel due to no pv response; please re-test.

The unit can be changed for the OFFSET axis, this is simply one casacore::Quantity conversion to the new unit.

I will write a short doc describing the fixed pixel vs angular separation.

pford commented:
Line box regions implementation is described in this document (in CARTA shared drive backend development folder):
https://docs.google.com/document/d/1VZqXM-A8Djmsybz4kNJsXudoghnjtqFvadB8SxlqEUQ/edit?usp=sharing

In both cases, the rectangle region's width is set to the width in the PvRequest message.  But this should be variable like the height for the angular separation case? (width is perpendicular to line, height is along line)

pford commented:
Converted this PR to draft while testing and tuning.  Still could use a code review of the basic strategy.

@kswang1029: I am testing a widefield image that uses the fixed angular increment which produces null box regions (casacore error = "LCPolygon - entire polygon is outside the lattice") between two valid regions, hence the NaN offsets.  Investigating this error.

pford committed:
Merged dev into branch

pford committed:
Merge branch 'dev' into pv_generator

pford committed:
Merge branch 'dev' into pv_generator

pford committed:
Update protobuf commit for updated changelog

pford committed:
Fix tbb merge error

pford committed:
Merge branch 'dev' into pv_generator

pford committed:
Update protobuf with dev

pford committed:
Improve debug output for polygon region approximation

pford committed:
Remove PvGenerator from Frame

pford commented:
Removing code reviewers while refactoring and until user testing is further along.

pford committed:
Move PvGenerator from Frame to RegionHandler

pford committed:
Expose spectral axis and image in Frame

pford committed:
Session start/cancel PV generator in  RegionHandler not Frame

pford committed:
Get spectral profile from temp region for pv image data in RegionHandler

pford committed:
Merge branch 'dev' into pv_generator

pford commented:
Pushed refactoring change - spectral profiles are calculated from each temporary box region rather than making a vector of box regions and getting profiles later.  Also added debug output; each rectangle region along the box is printed as "polygon[[...]]" in CRTF pixel format for pasting into a file to import and check, as well as max of NumPixels spectral profile for each region (always seems to be nan, but spectral profile is not nan).  This was not nan when spectral profiles were calculated in the PvGenerator, usually 0-3 pixels for width=3.

Preliminary testing looks okay.  I will continue testing and investigate the NumPixels issue.  Still need to implement fixed angular width.

pford committed:
Add NumPixels to list of spectral profile stats for pvgen testing

pford committed:
Refactor use polygon approximation function

pford committed:
Merge branch 'dev' into pv_generator

pford commented:
NumPixels = nan was because this stats type was not enabled for spectral profiles.  It is fixed now.

pford committed:
Remove debug output for pvgen from Region

pford committed:
Set regions with fixed angular spacing as polygons for pv image

pford committed:
Merge branch 'dev' into pv_generator

pford commented:
Fixed angular regions were changed to use polygon; will update implementation doc.  Ready for testing.  I will also do more testing on Monday.

kswang1029 commented:
> Fixed angular regions were changed to use polygon; will update implementation doc. Ready for testing. I will also do more testing on Monday.

I will start testing and report asap.

pford committed:
Revert comment

pford committed:
Merge branch 'dev' into pv_generator

pford committed:
Use pixel width to find line region corners

pford committed:
Shorten pv cut to line segment in image

Cannot convert endpoint pixels outside image to wcs.  Also calculate
positive offsets then negative offsets rather than alternate.

pford committed:
Fix format

pford committed:
Revert change shortening line cut to inside image

Also do not report errors of temporary regions outside image

pford committed:
Merge branch 'dev' into pv_generator

pford committed:
Implement overlapping polygons for line spectral profiles

Remove nan profiles for regions which contain no pixel centers

pford committed:
Fix crash when point vector varies in size

pford committed:
Merge branch 'dev' into pv_generator

pford committed:
Session stop pv in region handler without frame check

pford committed:
Improve PV cancellation

Add checks for stop flag, unset spectral requirements for
temp region to stop profile

pford committed:
Adjust units of offset axis in PvGenerator

pford committed:
Fix milliarcsecond unit - marcsec

pford commented:
Converted from draft PR at @kswang1029 's request for code review.  Most of the implementation is in RegionHandler, not sure who should be assigned to this PR. 

Remaining issue: when a line is matched between two widefield images and the PV image is requested for the matched image, the PV image is incorrect.  The converted line regions are the 1000-point polygon approximations, but this works when a single region is set and the z profile is requested for the matched region.  Still investigating this bug.

pford committed:
Use line region csys for setting line approximation regions

pford commented:
Remaining issue has been fixed (subject to further testing)

kswang1029 reviewed (approved):
Thank you @pford to bring the implementation of pv generator this far! 👍 I have tested some use cases including cancellation with small and large cubes, pv generation with narrow/wide field images, pv generation from matched narrow/wide field images etc. No obvious issues are seen for now. Hope we can learn more from users with the v3beta2 release.

kswang1029 commented:
@pford I think we need a changelog for this PR.

pford committed:
Add changelog entry

pford committed:
Update protobuf commit for changelog

pford commented:
I updated the backend and protobuf changelogs.  @kswang1029 let me know if I need to remove any debug output for the line regions (polygon points, num pixels).

kswang1029 commented:
> I updated the backend and protobuf changelogs. @kswang1029 let me know if I need to remove any debug output for the line regions (polygon points, num pixels).

Let’s keep it in the debug mode for now just in case. 🙂

pford committed:
Minor refactoring - comments, order of function parameters

pford committed:
Send pv progress at half second intervals

veggiesaurus commented:
@pford can you make a separate PR for the protobufs please?

veggiesaurus commented on src/ImageGenerators/PvGenerator.h:
```diff
@@ -0,0 +1,46 @@
+class PvGenerator {
+public:
+    PvGenerator(int file_id, const std::string& filename);
+    ~PvGenerator(){};
```
perhaps replace with `= default`?

pford replied:
Or just remove?  I think this is a refactoring relic.

I also think I don't need to have a map of pv generators in RegionHandler, since the cancellation is done in RegionHandler.  The pv generator is now a local variable after the profiles (pv data) are complete.

veggiesaurus commented:
@pford I don't have many comments about the code itself, it looks good. But could we add some basic unit tests for the PV generator? I'm happy to have them done in a separate PR if we want to get beta-2 out the door soon, but I do think this could benefit from some unit tests to catch potential regressions as we add functionality going forward.

pford committed:
Remove unneeded dtor and map

pford committed:
Merge branch 'dev' into pv_generator

pford committed:
Add PvGenerator test

pford commented:
I added TestPvGenerator, which uses test images in the backend repo (test/data/images/) since the generated images do not seem to have a valid coordinate system, which is required.  I check the PV image coordinate system, data shape, and that the data is not all NAN, and that an image with no spectral coordinate fails.

veggiesaurus commented:
> I added TestPvGenerator, which uses test images in the backend repo (test/data/images/) since the generated images do not seem to have a valid coordinate system, which is required. I check the PV image coordinate system, data shape, and that the data is not all NAN, and that an image with no spectral coordinate fails.

Ok. I think that's a good start. We should probably (a) adjust the generator to include a valid coord system and (b) add more robust tests in separate issues.

pford committed:
Set PV offset direction for horizontal line with fixed pixel spacing

pford committed:
Add PV test with horizontal line compared to image slice

pford committed:
Update protobuf commit and version

pford commented:
@veggiesaurus I thought I could easily add a PV generator test with a horizontal line and width 1, which would be equivalent to an image data slice with an x-range at the value of y.  I found that the PV data was the same no matter which way the line start/end was set, which turned out to be a bug in the offset direction for a horizontal line with fixed pixel spacing.  I should add a vertical line test as well.

I fixed the bug, added the test, and updated protobuf with the new commit and major version.  Will do more testing tomorrow, but @kswang1029 you might want to have another look at the pv image offset direction with fixed pixel increment if you have time.  The fixed angular offset direction should be correct since the separation is calculated toward the starting or ending endpoint.

pford commented:
Testing may have to wait until frontend ICD version is updated...

kswang1029 commented:
> @veggiesaurus I thought I could easily add a PV generator test with a horizontal line and width 1, which would be equivalent to an image data slice with an x-range at the value of y. I found that the PV data was the same no matter which way the line start/end was set, which turned out to be a bug in the offset direction for a horizontal line with fixed pixel spacing. I should add a vertical line test as well.
> 
> I fixed the bug, added the test, and updated protobuf with the new commit and major version. Will do more testing tomorrow, but @kswang1029 you might want to have another look at the pv image offset direction with fixed pixel increment if you have time. The fixed angular offset direction should be correct since the separation is calculated toward the starting or ending endpoint.

will do shortly and report back

pford committed:
Set height 3 for rotated fixed pixel line regions

pford committed:
Use float for pv line region height

pford committed:
Use ReadRegion test util and add pv vertical cut test

pford commented:
Created pv_generator branch in ICD tests to set IcdVersion = 25 and opened PR.

kswang1029 commented:
this commit 5ef8d36 helps a lot to remove the sampling artifacts. 
The middle image is from CARTA pv generator and the right image is from CASA-impv. It is hard to tell differences visually. Numerical values differ a bit but this is understandable as CASA-impv does image regrid and CARTA does box averaging.

![Screen Shot 2021-12-08 at 12 01 42 PM](https://user-images.githubusercontent.com/20819712/145146461-dadef92a-1f22-4f80-83d9-d0bf0fdf5d2e.png)

@veggiesaurus ready to merge in. @pford thank you again. This is one of the big milestones in the CARTA dev history.
